### PR TITLE
Make "is_required" and "is_visible" properties of telephone, company and fax attributes of addresses configurable

### LINKED ISSUE #1691: Telephone number on customer address shouldn't be required
Hi guys

Why is the telephone number still required on a customer address?
We hear from our shop owners that a lot of customers refuse to enter something in there or just enter gibberish.

An ideal solution would be if you could add a configuration option in the backend similar to the `Gender`, `Tax/Vat Number` and `Date of Birth` fields.
So an option for showing the telephone number where you can choose between `No`, `Optional` or `Required`. What do you think?

I had to implement this today into a Magento CE 1.x store and it was kind of a terrible experience, changing stuff in the database, remove the hard coded validation of that field (which is still present btw: https://github.com/magento/magento2/blob/93312cae42824274722fe7a285d5ce8e40f177a1/app/code/Magento/Customer/Model/Address/AbstractAddress.php#L572), ...

Related to https://github.com/magento/magento2/issues/624

### PR DESCRIPTION
**Pull request by avstudnitz**
Fixes #1691.
We added configuration for telephone, fax and company fields in configuration:
grafik
You can set them to be optional, required, or not shown at all. All necessary forms have been adjusted.

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=8519) <br/>All committers have signed the CLA.

File: app/code/Magento/Customer/view/frontend/templates/widget/company.phtml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+/** @var \Magento\Customer\Block\Widget\Company $block */
+?>
+
+<div class="field company<?php if ($block->isRequired()) echo ' required' ?>">
+    <label for="company" class="label"><span><?php echo $block->escapeHtml(__('Company')) ?></span></label>
+    <div class="control">
+        <input type="text" name="company" id="company" value="<?php echo $block->escapeHtmlAttr($block->getCompany()) ?>" title="<?php echo $block->escapeHtmlAttr(__('Company')) ?>" class="input-text <?php echo $block->escapeHtmlAttr($this->helper('Magento\Customer\Helper\Address')->getAttributeValidationClass('company')) ?>">
+    </div>
+</div>
```
Comment (Igloczek) [inline app/code/Magento/Customer/view/frontend/templates/widget/company.phtml]: Will be nice to improve code formatting in templates 😉  Copy-paste code bellow.

```phtml
<div class="field company <?php echo $block->isRequired() ? 'required' : '' ?>">
    <label for="company" class="label">
        <span>
            <?php echo $block->escapeHtml(__('Company')) ?>
        </span>
    </label>
    <div class="control">
        <?php
            $_validationClass = $block->escapeHtmlAttr(
                $this->helper('Magento\Customer\Helper\Address')
                     ->getAttributeValidationClass('company')
            );
        ?>
        <input type="text"
               name="company"
               id="company"
               value="<?php echo $block->escapeHtmlAttr($block->getCompany()) ?>"
               title="<?php echo $block->escapeHtmlAttr(__('Company')) ?>"
               class="input-text <?php echo $_validationClass ?? '' ?>"
         >
    </div>
</div>
```
PS. Remember about new line at the end of file.

File: app/code/Magento/Customer/view/frontend/templates/widget/fax.phtml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+/** @var \Magento\Customer\Block\Widget\Fax $block */
+?>
+
+<div class="field fax<?php if ($block->isRequired()) echo ' required' ?>">
+    <label for="fax" class="label"><span><?php echo $block->escapeHtml(__('Fax')) ?></span></label>
+    <div class="control">
+        <input type="text" name="fax" id="fax" value="<?php echo $block->escapeHtmlAttr($block->getFax()) ?>" title="<?php echo $block->escapeHtmlAttr(__('Fax')) ?>" class="input-text <?php echo $block->escapeHtmlAttr($this->helper('Magento\Customer\Helper\Address')->getAttributeValidationClass('fax')) ?>">
+    </div>
+</div>
```
Comment (Igloczek) [inline app/code/Magento/Customer/view/frontend/templates/widget/fax.phtml]:
```phtml
<div class="field fax <?php echo $block->isRequired() ? 'required' : '' ?>">
    <label for="fax" class="label">
        <span>
            <?php echo $block->escapeHtml(__('Fax')) ?>
        </span>
    </label>
    <div class="control">
        <?php
            $_validationClass = $block->escapeHtmlAttr(
                $this->helper('Magento\Customer\Helper\Address')
                     ->getAttributeValidationClass('fax')
            );
        ?>
        <input type="text"
               name="fax"
               id="fax"
               value="<?php echo $block->escapeHtmlAttr($block->getFax()) ?>"
               title="<?php echo $block->escapeHtmlAttr(__('Fax')) ?>"
               class="input-text <?php echo $_validationClass ?? '' ?>"
        >
    </div>
</div>
```

File: app/code/Magento/Customer/view/frontend/templates/widget/telephone.phtml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+/** @var \Magento\Customer\Block\Widget\Telephone $block */
+?>
+
+<div class="field telephone<?php if ($block->isRequired()) echo ' required' ?>">
+    <label for="telephone" class="label"><span><?php echo $block->escapeHtml(__('Phone Number')) ?></span></label>
+    <div class="control">
+        <input type="text" name="telephone" id="telephone" value="<?php echo $block->escapeHtmlAttr($block->getTelephone()) ?>" title="<?php echo $block->escapeHtmlAttr(__('Phone Number')) ?>" class="input-text <?php echo $block->escapeHtmlAttr($this->helper('Magento\Customer\Helper\Address')->getAttributeValidationClass('telephone')) ?>">
+    </div>
+</div>
```
Comment (Igloczek) [inline app/code/Magento/Customer/view/frontend/templates/widget/telephone.phtml]:
```phtml
<div class="field telephone <?php echo $block->isRequired() ? 'required' : '' ?>">
    <label for="telephone" class="label">
        <span>
            <?php echo $block->escapeHtml(__('Phone Number')) ?>
        </span>
    </label>
    <div class="control">
        <?php
            $_validationClass = $block->escapeHtmlAttr(
                $this->helper('Magento\Customer\Helper\Address')
                     ->getAttributeValidationClass('fax')
            );
        ?>
        <input type="text"
               name="telephone"
               id="telephone"
               value="<?php echo $block->escapeHtmlAttr($block->getTelephone()) ?>"
               title="<?php echo $block->escapeHtmlAttr(__('Phone Number')) ?>"
               class="input-text <?php echo $_validationClass ?? '' ?>"
        >
    </div>
</div>
```

**Review by Igloczek — CHANGES_REQUESTED** [review]

**Comment by hostep** [comment]
Nice work @avstudnitz!

Just something I have to think about, take a look at this code:
https://github.com/magento/magento2/blob/7c81c1dbeaea2cb7c3aed644580f48bbbdb226e9/app/code/Magento/Checkout/Block/Checkout/LayoutProcessor.php#L323-L339
I think you'll need to figure something out, to add the `required-entry` validation of those 3 billing address fields in there, based on the chosen configuration?

(not 100% sure about this though)

**Comment by avstudnitz** [comment]
@hostep In our tests it set the correct validation on the checkout without modifying anything, based on the attribute settings. so I think this is not necessary.

**Comment by fooman** [comment]
Nice work - I believe if you want to increase your chances of a quick merge I suggest taking a look at making the constructor changes here https://github.com/magento/magento2/pull/8519/files#diff-abecee6a6fc7dfb7ec46099d0ffec980R55 use an optional parameter similar to how @dmanners has been doing here https://github.com/magento/magento2/pull/8481/files#diff-ab7a66b8a18582a01b9661fe9270486aR51

**Comment by vrann** [comment]
@avstudnitz Thank you for the contribution! I'm verifying it at the moment.
Can you please take a look at the failing Travis tests?

File: app/code/Magento/Customer/Model/Address/AbstractAddress.php [hunk 4]
```diff
@@ -25,6 +25,8 @@
  * @method int getCountryId()
  * @method string getCity()
  * @method string getTelephone()
+ * @method string getCompany()
+ * @method string getFax()
```
Comment (vrann) [inline app/code/Magento/Customer/Model/Address/AbstractAddress.php:29]: nice!

**Comment by avstudnitz** [comment]
@vrann Yes, will do that over the weekend. It's the unit tests, right?

**Comment by vrann** [comment]
@avstudnitz actually nevermind issues in the unit tests (Travis), it was fast fix and I just did it.
Verifying the functionality and will let you know is there anything need to be changed.

**Comment by vrann** [comment]
@avstudnitz from the review perspective all the changes looks good and the functionality works perfectly.

**Comment by avstudnitz** [comment]
Great to hear @vrann!

**Comment by okorshenko** [comment]
@avstudnitz thank you for your contribution. PR is merged

**Pull request merged**